### PR TITLE
Check for serial connection lost

### PR DESCRIPTION
**Pull request by microbit-grace**

### CODE REVIEW

File: src/script/microbit-interfacing/MicrobitSerial.ts [hunk 1]
```diff
@@ -104,26 +130,54 @@ export class MicrobitSerial implements MicrobitConnection {
 
       stateOnAssigned(DeviceRequestStates.INPUT, this.usb.getModelNumber());
       stateOnReady(DeviceRequestStates.INPUT);
+      logMessage('Serial successfully connected');
     } catch (e) {
       logError('Failed to initialise serial protocol', e);
       stateOnFailedToConnect(DeviceRequestStates.INPUT);
       await this.usb.stopSerial();
       throw e;
```
Comment (microbit-grace) [inline src/script/microbit-interfacing/MicrobitSerial.ts:138]: Interesting.. for some reasons the first reconnect after connection loss (clicking reset btn on micro:bit) currently fail. The error message includes: `Error: Timeout waiting for response 530415182`

Comment (microbit-grace) [reply src/script/microbit-interfacing/MicrobitSerial.ts:138]: With help from Matt, this seems to be fixed by stopping serial before trying to reconnect

File: src/script/microbit-interfacing/MicrobitSerial.ts [hunk 2]
```diff
@@ -80,6 +94,19 @@ export class MicrobitSerial implements MicrobitConnection {
       await this.handshake();
       stateOnConnected(DeviceRequestStates.INPUT);
 
+      // Check for connection lost
+      if (this.connectionCheckIntervalId === undefined) {
+        this.connectionCheckIntervalId = setInterval(async () => {
+          const allowedTimeWithoutMessageInMs = isDevMode ? 2000 : 10000;
```
Comment (microbit-robert) [inline src/script/microbit-interfacing/MicrobitSerial.ts:100]: We can use StaticConfiguration.connectTimeoutDuration for the 10 second timeout to match Bluetooth.

File: src/script/microbit-interfacing/MicrobitSerial.ts [hunk 3]
```diff
@@ -104,26 +131,55 @@ export class MicrobitSerial implements MicrobitConnection {
...
+  async handleReconnect(): Promise<void> {
+    if (this.isConnecting) {
+      logMessage('Serial disconnect ignored... reconnect already in progress');
+      return;
+    }
+    try {
+      this.stopConnectionCheck();
+      logMessage('Serial disconnected... automatically trying to reconnect');
+      await this.usb.softwareReset();
+      await this.usb.stopSerial();
```
[the remark this answers is not in the document]
Comment (microbit-grace) [reply src/script/microbit-interfacing/MicrobitSerial.ts:172]: I was really stuck and tried both. Both throw an error 😢

File: src/script/microbit-interfacing/MicrobitSerial.ts (hunk 2 above)
Comment (microbit-grace) [reply src/script/microbit-interfacing/MicrobitSerial.ts:100]: Nice, didn't know that existed

**Comment by microbit-robert** [comment]
Resetting the bridge:
image

Turning off the remote:
image

**Review by microbit-robert — APPROVED** [review]
LGTM.

**Pull request merged**